src/test/java/codeu/controller/AdminServletTest.java

```diff
@@ -63,27 +64,51 @@ public void testDoGet_NullUser() throws IOException, ServletException {
     Mockito.verify(mockResponse).sendRedirect("/login");
   }
 
+  @Test
+  public void testDoGet_FakeUser() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("fake_username");
+    Mockito.when(mockUserStore.getUser("fake_username")).thenReturn(null);
+
+    adminServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockResponse).sendRedirect("/login");
+  }
+
 /*
   @Test
   public void testDoGet_ValidUser() throws IOException, ServletException {
+    User admin_user =
+        new User(
+            UUID.randomUUID(),
+            "admin_username",
+            "admin_password",
+            Instant.now());
+
     Mockito.when(mockSession.getAttribute("user")).thenReturn("admin_username");
+    Mockito.when(mockUserStore.getUser("admin_username")).thenReturn(admin_user);
 
     adminServlet.doGet(mockRequest, mockResponse);
 
-    Mockito.verify(mockSession).getAttribute("user").matches("admin_username");
     Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
   }
-
+  */
 
   @Test
-  public void testDoGet_NotValidUser() throws IOException, ServletException {
-    Mockito.when(mockSession.getAttribute("user")).thenReturn("admin_username");
+  public void testDoGet_NotAdminUser() throws IOException, ServletException {
+    User not_admin_user =
+        new User(
+            UUID.randomUUID(),
+            "not_admin_username",
+            "not_admin_password",
+            Instant.now());
+
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("not_admin_username");
+    Mockito.when(mockUserStore.getUser("not_admin_username")).thenReturn(not_admin_user);
 
     adminServlet.doGet(mockRequest, mockResponse);
 
-    Mockito.verify((mockSession).getAttribute("user"), never()).matches("admin_username");
     Mockito.verify(mockResponse).sendRedirect("/");
```

Comment:
So, the point of this test is to demonstrate that any user other than the Chosen Few are redirected away from the admin page, correct? That's pretty easy to see from studying the code, but for someone unfamiliar it would save time to have at least a short comment to that effect. It's good your method has a descriptive name.
I presume you'll add a positive test for the 3 admins?

Reply:
Any longer term ideas for identifying admins other than the hard code list?

Reply:
- I will add a comment to describe what the test does!
- Yes, I was working on making it but was getting a mistake (so I left it commented) I'll make it work as the days pass by.
- Yeah, I was thinking of adding the admins to an unordered set (hashset in java) that gives me constant O(1) average lookup and addition time complexity, in exchange of some memory O(admins). What do you think?

Reply:
-Another great option that is better for scalability (I think so) is adding an attribute to each User, a boolean telling us if he's or not admin. This would be great for the future, in case I wanted to show some admin features in every .jsp!

Reply:
I like that last idea. I'm not very aware of data tables like has tables (so I have no idea which scales better), but this idea sound great to me.